### PR TITLE
Add support for postgres 11, add more postgres version, factor files and add a generator

### PR DESCRIPTION
**Pull request by desaintmartin**

### CODE REVIEW

**Comment by desaintmartin** [comment]
Do you have any interest into going forward with this idea? Should I just close it?

**Comment by Vampouille** [comment]
> Do you have any interest into going forward with this idea? Should I just close it?

Yes it's interesting to improve the maintainability of this repo, currently there is a lot of code duplications.
I think it will be easier to use build time parameter and use this variable in Dockerfiles. This way we can have only one Dockerfile for 9.x and another for 1x.x. Docker hub supports build argument.

https://docs.docker.com/engine/reference/commandline/#set-build-time-variables---build-arg

If you're interested in restoration ;-) I added an entrypoint dedicated to restoration that populates the volume for postgres.
https://github.com/camptocamp/docker-pghoard/blob/test/10/restore-entrypoint.sh

```yaml
version: '2'
services:
  pghoard-restore:
    image: camptocamp/pghoard:10-test
    entrypoint: "/restore-entrypoint.sh"
    command: "test timestamp 2019-09-10T16:59:00+02:00 true promote /var/lib/postgresql/data"
    env_file: pghoard_config.env
    environment:
      PGHOARD_CONFIG: /mnt/config/pghoard.json
    volumes:
    - pgdata-restore:/var/lib/postgresql/data
    - pghoard-config:/mnt/config
  postgres-restore:
    image: camptocamp/pghoard:10-test
    entrypoint: "/usr/local/bin/wait-for-it.sh"
    command: "--host=pghoard --port=16000  --timeout=0 --strict -- /usr/local/bin/docker-entrypoint.sh postgres"
    links:
      - pghoard-restore:pghoard
    environment:
      POSTGRES_PASSWORD: pgpass
      POSTGRES_USER: postgres
    volumes:
      - pgdata-restore:/var/lib/postgresql/data
volumes:
  pgdata-restore: {}

```

**Comment by desaintmartin** [comment]
Don't you want to auto-build from docker hub? I am not aware of additional params you can pass there.
Any way I am still interested into factoring all of this. :)

Edit: yes it does, let me try.

**Comment by desaintmartin** [comment]
What do you think about this approach (forgetting the restore for now)?

File: README.md [hunk 1]
```diff
@@ -88,28 +95,14 @@ Can be either `telegraf` or `datadog`. (default `telegraf`)
 
 ## OS_TENANT_NAME
 
-## ENCRYPTION_KEYS_PUBLIC
+## PGHOARD_RESTORE_SITE
```
Comment (Vampouille) [inline README.md]: Please restore the documentation about `ENCRYPTION_KEYS_*`

File: restore.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
```
Comment (Vampouille) [inline restore.sh]: Can you move the restore part into another PR ?

File: make.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
+
+# This script will generate the Dockerfiles
```
Comment (Vampouille) [inline make.sh]: This script will build the images and push if an argument is provided

File: make.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+if [ -n "$1" ]; then
+  DOCKER_REPOSITORY="$1:"
+fi
+
+cd $(dirname $0)
+
+for VERSION in "9.5" "9.6" "10" "11" "12"; do
+    DOCKER_TAG_DATE=$DOCKER_REPOSITORY$VERSION-$(date +%Y-%m-%d)
+    DOCKER_TAG=$DOCKER_REPOSITORY$VERSION-$(date +%Y-%m-%d)
+    docker build --file Dockerfile --build-arg DOCKER_POSTGRES_VERSION="$VERSION" --tag "$DOCKER_TAG" --tag "$DOCKER_TAG_DATE" .
```
Comment (Vampouille) [inline make.sh]: you can remove the `--file` option as `Dockerfile` is the default value

File: make.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+if [ -n "$1" ]; then
+  DOCKER_REPOSITORY="$1:"
+fi
+
+cd $(dirname $0)
+
+for VERSION in "9.5" "9.6" "10" "11" "12"; do
+    DOCKER_TAG_DATE=$DOCKER_REPOSITORY$VERSION-$(date +%Y-%m-%d)
+    DOCKER_TAG=$DOCKER_REPOSITORY$VERSION-$(date +%Y-%m-%d)
```
Comment (Vampouille) [inline make.sh]: You have the same value for both tags, maybe a useful value for `DOCKER_TAG` can be `$DOCKER_REPOSITORY$VERSION` so you have 2 tags:
* camptocamp/pghoard:10-2019-10-30
* camptocamp/pghoard:10

File: docker-entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
```
Comment (Vampouille) [inline docker-entrypoint.sh]: Why a new file? You should keep the original entrypoint in this PR

File: templates/pghoard.json.tmpl [hunk 7]
```diff
@@ -12,24 +12,13 @@
          "basebackup_mode": "pipe",
          "nodes":[
             {
```
Comment (Vampouille) [inline templates/pghoard.json.tmpl]: Please keep the original file here and move changes to another PR

File: templates/pghoard_restore.json.tmpl [hunk 8]
```diff
@@ -0,0 +1,50 @@
+{
```
Comment (Vampouille) [inline templates/pghoard_restore.json.tmpl]: Please move this file to a "restore" PR

File: conf.d/pghoard_restore.json.toml [hunk 9]
```diff
@@ -0,0 +1,3 @@
+[template]
```
Comment (Vampouille) [inline conf.d/pghoard_restore.json.toml]: Please move this in a "restore" PR

File: Dockerfile [hunk 10]
```diff
@@ -69,6 +71,7 @@ COPY conf.d /etc/confd/conf.d
 COPY templates /etc/confd/templates
 
 COPY /docker-entrypoint.sh /
+COPY /restore.sh /
```
Comment (Vampouille) [inline Dockerfile]: Please move this to a "restore" PR

File: Dockerfile [hunk 11]
```diff
@@ -1,4 +1,6 @@
-FROM postgres:10
+ARG DOCKER_POSTGRES_VERSION=10
+
+FROM postgres:$DOCKER_POSTGRES_VERSION
```
Comment (Vampouille) [inline Dockerfile]: It would be nice if you can also add variable for the base image name:
```Dockerfile
 ARG DOCKER_POSTGRES_IMAGE=postgres
```
because we would like to use the `camptocamp/postgres` as base image to be able to restore databases with postgis extension.

**Review by Vampouille — CHANGES_REQUESTED** [review]
* Please move restore stuff in another PR
* Restore entrypoint and stuff about encryption

**Comment by Vampouille** [comment]
> What do you think about this approach (forgetting the restore for now)?

It's a good approach, I think we can merge this if you address comments (see review). Thanks for this contribution !

File: README.md (hunk 1 above)
Comment (desaintmartin) [reply README.md]: I must have messed up something with the diffs. I'll restore all that was there before.

File: Dockerfile (hunk 11 above)
Comment (desaintmartin) [reply Dockerfile]: Done!

File: make.sh (hunk 3 above)
Comment (desaintmartin) [reply make.sh]: Done

File: make.sh (hunk 5 above)
Comment (desaintmartin) [reply make.sh]: Yes, my bad.

File: make.sh (hunk 4 above)
Comment (desaintmartin) [reply make.sh]: Done.

File: restore.sh (hunk 2 above)
Comment (desaintmartin) [reply restore.sh]: Sure!

**Comment by desaintmartin** [comment]
I've cleaned-up the PR, it should be much better now.
Note that postgresql 12 does not work with the version of pghoard set here, but using a newer commit makes it work.
